sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,770 @@
+        # Start all points as 'unprocessed' ##
+        self._processed = np.zeros((n_samples, 1), dtype=bool)
+        self.reachability_ = np.ones(n_samples) * np.inf
+        self.core_dists_ = np.ones(n_samples) * np.nan
+        # Start all points as noise ##
+        self.labels_ = -np.ones(n_samples, dtype=int)
+        self.ordering_ = []
+
+        # Check for valid n_samples relative to min_samples
+        if self.min_samples > n_samples:
+            raise ValueError("Number of training samples (n_samples=%d) must "
+                             "be greater than min_samples (min_samples=%d) "
+                             "used for clustering." %
+                             (n_samples, self.min_samples))
+
+        super(OPTICS, self).fit(X)
+
+        self.core_dists_[:] = self.kneighbors(X, self.min_samples)[0][:, -1]
+
+        # Main OPTICS loop. Not parallelizable. The order that entries are
+        # written to the 'ordering_' list is important!
+        for point in range(n_samples):
+            if not self._processed[point]:
+                self._expand_cluster_order(point, X)
+
+        indices_, self.labels_ = _extract_optics(self.ordering_,
+                                                 self.reachability_,
+                                                 self.maxima_ratio,
+                                                 self.rejection_ratio,
+                                                 self.similarity_threshold,
+                                                 self.significant_min,
+                                                 self.min_cluster_size_ratio,
+                                                 self.min_maxima_ratio)
+        self.core_sample_indices_ = indices_
+        self.n_clusters_ = np.max(self.labels_)
+        return self
+
+    # OPTICS helper functions; these should not be public #
+
+    def _expand_cluster_order(self, point, X):
+        # As above, not parallelizable. Parallelizing would allow items in
+        # the 'unprocessed' list to switch to 'processed'
+        if self.core_dists_[point] <= self.max_bound:
+            while not self._processed[point]:
+                self._processed[point] = True
+                self.ordering_.append(point)
+                point = self._set_reach_dist(point, X)
+        else:
+            self.ordering_.append(point)  # For very noisy points
+            self._processed[point] = True
+
+    def _set_reach_dist(self, point_index, X):
+        P = np.array(X[point_index]).reshape(1, -1)
+        indices = self.radius_neighbors(P, radius=self.max_bound,
+                                        return_distance=False)[0]
+
+        # Checks to see if there more than one member in the neighborhood
+        if iterable(indices):
+            # Masking processed values; n_pr is 'not processed'
+            n_pr = np.compress((np.take(self._processed,
+                                        indices, axis=0) < 1).ravel(),
```

Comment:
avoid axis=0 where the input is 1d, please

Reply:
fixed.